### ISSUE TITLE
BuildKonfig `0.9.0` -> `0.10.0` breaks Gradle Sync

### ISSUE THREAD

**Issue by chris-hatton** [yshrsmz/BuildKonfig#56]
Thanks for **BuildKonfig**, it continues to be an essential tool for KMP App Development! ❤️ 
I was happy to see continued maintenance with a `0.10.0` release:

Unfortunately, this new version has a bug for my project.
Gradle Sync fails with the error:
```
Duplicate content roots detected:
Path [/Users/me/project/client/build/buildkonfig/commonMain] of module 
[project.client.commonMain] was removed from modules [project.client.materialComposeMain]
```
_(I've redacted my actual username and project name but the pattern is preserved)_
This prevents working with the project.

If I change only the **BuildKonfig** dependency back to `0.9.0`, everything works as expected.

My project has quite a nested module structure; perhaps you've changed some code related to adding source-set paths to modules that has caused it to add to multiple source-sets unnecessarily?

**Comment by yshrsmz** [comment]
Hi,

Indeed I changed the source-sets registration logic in 0.10.0.

[`v0.9.0...v0.10.1`#diff-dd276d30d4](https://github.com/yshrsmz/BuildKonfig/compare/v0.9.0...v0.10.1#diff-dd276d30d4f21d6e0aefb7855c6711d86a1f4a29eb2f0a7ad80e7798ea07e2f6)

Previously the plugin explicitly obtained `commonMain` source-sets and registered buildkonfig dependency, but now it adds the dependency when the target is `KotlinMetadataTarget` which is common source-sets.

I thought there's only one `KotlinMetadataTarget` in a single KMP module, but looks like it was wrong.

Can you provide your targets configuration in `client` module's `build.gradle`?
I'd like to see how each source-sets are depending to each other.

(Thanks for sponsoring me btw! It means a lot to me.)

**Comment by chris-hatton** [comment]
Sure, my source sets dependencies (this is within a **single module**) look like this:

![image](https://user-images.githubusercontent.com/895683/135465765-0a1aa2ea-97ec-49e8-ac6e-c85a638976f4.png)

So far in KMP it has probably been less common to have these intermediary source-sets; but it will certainly become more common in the near future, as Compose/Multiplatform gains more use (this is what I am using now).

Also, JetBrains do show some examples of setting up source-sets like this in their KMP introductory material; just to show off the possibilities.  [See the source-set configurations discussed here](https://kotlinlang.org/docs/mpp-share-on-platforms.html).

**Comment by yshrsmz** [comment]
Here's what I found so far.

Actually, there's only one KotlinMetadataTarget, but there are multiple `main` compilations in it.


When `kotlin.mpp.enableGranularSourceSetsMetadata` is set to true, KMP adds dedicated compilation for shared sourceset in KotlinMetadataTarget

so

```
kotlin {
    android()
    jvm {}
    js(IR) {
        browser()
    }
    js("node") {
        nodejs()
    }
    macosX64()
    linuxX64()
    mingwX64()

    sourceSets {
        val commonMain by getting
        val commonTest by getting {
            dependencies {
                implementation(kotlin("test"))
            }
        }
        val jvmMain by getting
        val jvmTest by getting

        val appMain by creating {
            dependsOn(commonMain)
        }

        val androidMain by getting {
            dependsOn(appMain)
        }

        val jsCommonMain by creating {
            dependsOn(commonMain)
        }
        val jsMain by getting {
            dependsOn(jsCommonMain)
        }
        val nodeMain by getting {
            dependsOn(jsCommonMain)
        }
        val jsTest by getting

        val desktopMain by creating {
            dependsOn(appMain)
        }
        val macosX64Main by getting {
            dependsOn(desktopMain)
        }
        val linuxX64Main by getting {
            dependsOn(desktopMain)
        }
        val mingwX64Main by getting {
            dependsOn(desktopMain)
        }
    }
}
```

so if we have a source set configuration above, KotlinMetadataTarget has compilations below

```
compilation: appMain, class org.jetbrains.kotlin.gradle.plugin.mpp.KotlinCommonCompilation, task ':compileAppMainKotlinMetadata', metadataAppMainClasses
compilation: defaultSourceSet: source set appMain
compilation: allKotlinSourceSets: []
------
compilation: commonMain, class org.jetbrains.kotlin.gradle.plugin.mpp.KotlinCommonCompilation, task ':compileCommonMainKotlinMetadata', metadataCommonMainClasses
compilation: defaultSourceSet: source set commonMain
compilation: allKotlinSourceSets: []
------
compilation: desktopMain, class org.jetbrains.kotlin.gradle.plugin.mpp.KotlinSharedNativeCompilation, task ':compileDesktopMainKotlinMetadata', metadataDesktopMainClasses
compilation: defaultSourceSet: source set desktopMain
compilation: allKotlinSourceSets: []
------
compilation: main, class org.jetbrains.kotlin.gradle.plugin.mpp.KotlinCommonCompilation, task ':compileKotlinMetadata', metadataMainClasses
compilation: defaultSourceSet: source set commonMain
compilation: allKotlinSourceSets: [source set commonMain]
```

Looks like what we want is commonMain or main compilation, but I'm not sure which is the one.

But when I create a simple project without shared source set, there's one single main compilation

```
target: metadata, class org.jetbrains.kotlin.gradle.plugin.mpp.KotlinMetadataTarget_Decorated
compilation: main, class org.jetbrains.kotlin.gradle.plugin.mpp.KotlinCommonCompilation, task ':presentation:presentation-all:compileKotlinMetadata', metadataMainClasses
compilation: defaultSourceSet: source set commonMain
compilation: allKotlinSourceSets: [source set commonMain]
```

So for now I'm going to check the name of the compilation

**Comment by yshrsmz** [comment]
```
target: metadata, class org.jetbrains.kotlin.gradle.plugin.mpp.KotlinMetadataTarget_Decorated
compilation: appMain, class org.jetbrains.kotlin.gradle.plugin.mpp.KotlinCommonCompilation, task ':compileAppMainKotlinMetadata', metadataAppMainClasses
compilation: defaultSourceSet: source set appMain, dependsOn: [source set commonMain]
compilation: allKotlinSourceSets: []
------
compilation: commonMain, class org.jetbrains.kotlin.gradle.plugin.mpp.KotlinCommonCompilation, task ':compileCommonMainKotlinMetadata', metadataCommonMainClasses
compilation: defaultSourceSet: source set commonMain, dependsOn: []
compilation: allKotlinSourceSets: []
------
compilation: desktopMain, class org.jetbrains.kotlin.gradle.plugin.mpp.KotlinSharedNativeCompilation, task ':compileDesktopMainKotlinMetadata', metadataDesktopMainClasses
compilation: defaultSourceSet: source set desktopMain, dependsOn: [source set appMain]
compilation: allKotlinSourceSets: []
------
compilation: main, class org.jetbrains.kotlin.gradle.plugin.mpp.KotlinCommonCompilation, task ':compileKotlinMetadata', metadataMainClasses
compilation: defaultSourceSet: source set commonMain, dependsOn: []
compilation: allKotlinSourceSets: [source set commonMain]
------

Ok you are right.
I think we can just check defaultSourceSet dependency

**Comment by yshrsmz** [comment]
@chris-hatton 

Hi, I've just released v0.10.2 with a possible fix for this issue.
Please let me know if you find anything weird. thanks!
